File: step_06_linked_list/6.3_medium_problems/IB064_even_reverse/solution.py
"""Problem: Even Reverse - Reverse even-positioned nodes in a linked list (1-indexed)"""

# ============================================================
# ListNode definition
# ============================================================
class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt

    @staticmethod
    def from_list(vals):
        dummy = ListNode()
        cur = dummy
        for v in vals:
            cur.next = ListNode(v)
            cur = cur.next
        return dummy.next

    @staticmethod
    def to_list(head):
        result = []
        while head:
            result.append(head.val)
            head = head.next
        return result


# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(n)  |  Space: O(n)
# Collect values at even positions (1-indexed), reverse them, write back
# ============================================================
def brute_force(head: ListNode) -> ListNode:
    even_vals = []
    cur, pos = head, 1
    while cur:
        if pos % 2 == 0:
            even_vals.append(cur.val)
        cur = cur.next
        pos += 1
    even_vals.reverse()
    cur, pos, idx = head, 1, 0
    while cur:
        if pos % 2 == 0:
            cur.val = even_vals[idx]
            idx += 1
        cur = cur.next
        pos += 1
    return head


# ============================================================
# APPROACH 2: OPTIMAL - Extract even nodes, reverse sub-list, re-weave
# Time: O(n)  |  Space: O(n)
# Separate odd and even position nodes into two lists, reverse the even list,
# then merge them back alternately.
# ============================================================
def optimal(head: ListNode) -> ListNode:
    def reverse_list(node):
        prev, cur = None, node
        while cur:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        return prev

    odd_dummy, even_dummy = ListNode(), ListNode()
    odd_cur, even_cur = odd_dummy, even_dummy
    cur, pos = head, 1
    while cur:
        nxt = cur.next
        cur.next = None
        if pos % 2 == 1:
            odd_cur.next = cur
            odd_cur = odd_cur.next
        else:
            even_cur.next = cur
            even_cur = even_cur.next
        cur = nxt
        pos += 1

    even_head = reverse_list(even_dummy.next)
    # Interleave
    dummy = ListNode()
    res = dummy
    o, e = odd_dummy.next, even_head
    while o or e:
        if o:
            res.next = o
            res = res.next
            o = o.next
        if e:
            res.next = e
            res = res.next
            e = e.next
    res.next = None
    return dummy.next


# ============================================================
# APPROACH 3: BEST - Clean value-only approach (most Pythonic)
# Time: O(n)  |  Space: O(n/2)
# Same as brute force but cleaner with list slice reverse
# ============================================================
def best(head: ListNode) -> ListNode:
    # Collect nodes at even positions
    even_nodes = []
    cur, pos = head, 1
    while cur:
        if pos % 2 == 0:
            even_nodes.append(cur)
        cur = cur.next
        pos += 1
    # Reverse values in even-position nodes
    vals = [n.val for n in even_nodes]
    vals.reverse()
    for node, val in zip(even_nodes, vals):
        node.val = val
    return head


if __name__ == "__main__":
    print("=== Even Reverse ===")
    # 1->2->3->4->5: even positions are 2,4 -> reversed: 4,2 -> result: 1->4->3->2->5
    tests = [
        ([1, 2, 3, 4, 5], [1, 4, 3, 2, 5]),
        ([1, 2, 3, 4],    [1, 4, 3, 2]),
        ([1],              [1]),
        ([1, 2],           [1, 2]),
    ]
    for vals, expected in tests:
        b  = ListNode.to_list(brute_force(ListNode.from_list(vals)))
        o  = ListNode.to_list(optimal(ListNode.from_list(vals)))
        be = ListNode.to_list(best(ListNode.from_list(vals)))
        print(f"Input={vals} => Brute={b}, Optimal={o}, Best={be} (exp={expected})")
        assert b == o == be == expected, f"Mismatch: {b} {o} {be}"
    print("All tests passed.")

File: step_06_linked_list/6.3_medium_problems/IB064_even_reverse/Solution.java
import java.util.*;

// ============================================================
// ListNode definition
// ============================================================
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) { this.val = val; }

    static ListNode of(int... vals) {
        ListNode dummy = new ListNode(0), cur = dummy;
        for (int v : vals) { cur.next = new ListNode(v); cur = cur.next; }
        return dummy.next;
    }

    static String toStr(ListNode head) {
        StringBuilder sb = new StringBuilder("[");
        for (ListNode n = head; n != null; n = n.next) {
            sb.append(n.val); if (n.next != null) sb.append("->");
        }
        return sb.append("]").toString();
    }
}

// ============================================================
// APPROACH 1: BRUTE FORCE
// Time: O(n)  |  Space: O(n)
// Extract values at even positions (1-indexed: 2,4,6,...) into an array,
// reverse that array, then write values back.
// ============================================================
class BruteForce {
    public static ListNode solve(ListNode head) {
        // Collect even-position (1-indexed) values
        List<Integer> evenVals = new ArrayList<>();
        int pos = 1;
        for (ListNode n = head; n != null; n = n.next) {
            if (pos % 2 == 0) evenVals.add(n.val);
            pos++;
        }
        Collections.reverse(evenVals);
        // Write back
        int idx = 0; pos = 1;
        for (ListNode n = head; n != null; n = n.next) {
            if (pos % 2 == 0) { n.val = evenVals.get(idx++); }
            pos++;
        }
        return head;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - Extract even-position nodes, reverse, re-weave
// Time: O(n)  |  Space: O(n) for the sub-list
// Collect even-indexed nodes, reverse that sub-list in-place,
// then interleave back into the original list.
// ============================================================
class Optimal {
    private static ListNode reverse(ListNode head) {
        ListNode prev = null, cur = head;
        while (cur != null) { ListNode next = cur.next; cur.next = prev; prev = cur; cur = next; }
        return prev;
    }

    public static ListNode solve(ListNode head) {
        // Separate into odd and even positioned nodes (1-indexed)
        ListNode oddDummy  = new ListNode(0), oddCur  = oddDummy;
        ListNode evenDummy = new ListNode(0), evenCur = evenDummy;
        ListNode cur = head;
        int pos = 1;
        while (cur != null) {
            ListNode next = cur.next; cur.next = null;
            if (pos % 2 == 1) { oddCur.next = cur; oddCur = oddCur.next; }
            else              { evenCur.next = cur; evenCur = evenCur.next; }
            cur = next; pos++;
        }
        // Reverse even list
        ListNode evenHead = reverse(evenDummy.next);
        // Interleave odd and (reversed) even
        ListNode dummy = new ListNode(0), res = dummy;
        ListNode o = oddDummy.next, e = evenHead;
        while (o != null || e != null) {
            if (o != null) { res.next = o; res = res.next; o = o.next; }
            if (e != null) { res.next = e; res = res.next; e = e.next; }
        }
        return dummy.next;
    }
}

// ============================================================
// APPROACH 3: BEST - In-place value swap (avoid re-weaving)
// Time: O(n)  |  Space: O(n/2) for even values array
// Collect only even-position values, reverse, write back — clean O(n) pass
// ============================================================
class Best {
    public static ListNode solve(ListNode head) {
        // Count even-position nodes first
        List<Integer> evenVals = new ArrayList<>();
        int pos = 1;
        for (ListNode n = head; n != null; n = n.next) {
            if (pos % 2 == 0) evenVals.add(n.val);
            pos++;
        }
        if (evenVals.isEmpty()) return head;
        Collections.reverse(evenVals);
        int idx = 0; pos = 1;
        for (ListNode n = head; n != null; n = n.next) {
            if (pos % 2 == 0) n.val = evenVals.get(idx++);
            pos++;
        }
        return head;
    }
}

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Even Reverse ===");
        // 1->2->3->4->5  even positions (1-indexed): 2,4  reversed: 4,2  => 1->4->3->2->5
        int[][] tests = {
            {1, 2, 3, 4, 5},
            {1, 2, 3, 4},
            {1},
            {1, 2},
        };
        for (int[] vals : tests) {
            System.out.printf("Input: %s => Brute: %s  Optimal: %s  Best: %s%n",
                ListNode.toStr(ListNode.of(vals)),
                ListNode.toStr(BruteForce.solve(ListNode.of(vals))),
                ListNode.toStr(Optimal.solve(ListNode.of(vals))),
                ListNode.toStr(Best.solve(ListNode.of(vals))));
        }
    }
}
